src/LanguageServerProtocol/LanguageServerProtocol.fs

```diff
@@ -2399,36 +2458,98 @@ module Server =
             }
             loop ())
 
-        /// When the server wants to send a request/notification to the client
-        let sendServerRequest (rpcMethod: string) (requestObj: obj): AsyncLspResult<unit> =
-            let serializedResponse = JToken.FromObject(requestObj, jsonSerializer)
-            let req = JsonRpc.Request.Create(rpcMethod, serializedResponse)
-            let reqString = JsonConvert.SerializeObject(req, jsonSettings)
-            sender.Post(reqString)
-            // TODO: Really wait for the client answer if not a notification (Necessary to implement requests)
+        let responseAgent = MailboxProcessor<ResponseMailboxMsg>.Start(fun agent ->
+          let rec loop state =
+              async{
+                  let! msg = agent.Receive()
+                  match msg with
+                  | Request (rid, reply) ->
+                      return! loop (state |> Map.add rid reply)
+                  | Response (Some rid, value) ->
+                      let result = state |> Map.tryFind rid
+                      match result with
+                      |Some(reply) ->
+                          reply.Reply(value.Result)
+                      |None -> eprintfn "Unexpected response %i" rid
```

Comment:
How should I log in this project?

Reply:
Add a project reference to FsAutoComplete.Logging.fsproj and then open FsAutoComplete.Logging and get a logger using `LogProvider.getLoggerByName` (or the other factory functions). Then emit structured logs in the same way we do in other places (see the .Lsp.fs layer for a good set of examples)

Reply:
I'm not sure if I want to add a dependency to FsAutoComplete.Logging. 
In the best scenario `LanguageServerProtocol.fsproj` should be standalone, and releasable as a NuGet package without FSAC dependency

Reply:
Breaking it away is still very easy:

* in the new project, add the paket github dependency on the logger
* do the namespace replacement to a different namespace
* in this file change the open statement to this new namespace

Reply:
I've added logging where I think it's important, just using `FsAutoComplete.Logging`
The things you're mentioning are changes to the build process right?

Reply:
Yeah, those are the steps that would be necessary if someone were to split out the LSP project into a standalone repository. They were meant as a counter argument to @Krzysztof-Cieslak's point about wanting this project to remain 'pure'.